datafusion/core/src/dataframe/mod.rs

```diff
@@ -2498,6 +2513,34 @@ impl DataFrame {
         let df = ctx.read_batch(batch)?;
         Ok(df)
     }
+
+    /// Resolves the extension types in an expression, returning the same expression but with
+    /// resolved extension types. This is done by consulting the [`SessionState`].
+    ///
+    /// Only *unresolved* expressions will be resolved. If users already set the logical type
+    /// reference of an expression, this method does not check whether this is the "right" extension
+    /// type registered in the session.
+    fn resolve_types_expr(&self, expr: Expr) -> Result<Transformed<Expr>> {
```

Comment:
Expressions passed into the `DataFrame` will be resolved. We also need a similar handling for LogicalPlan. Maybe once we create a DataFrame/Stream/etc. from the `LogicalPlan` we resolve it?